hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/types/HazelcastTypeCoercion.java

```diff
@@ -0,0 +1,422 @@
+        // Infer common type from columns and sub-expressions.
+
+        RelDataType commonType = null;
+        boolean seenParameters = false;
+        boolean seenChar = false;
+
+        for (SqlNode operand : operands) {
+            RelDataType operandType = validator.deriveType(scope, operand);
+            if (isLiteral(operand)) {
+                continue;
+            }
+
+            if (isParameter(operand)) {
+                seenParameters = true;
+            } else {
+                commonType = commonType == null ? operandType : withHigherPrecedence(operandType, commonType);
+                seenChar |= isChar(operandType);
+            }
+        }
+
+        // Continue common type inference on numeric literals.
+
+        for (SqlNode operand : operands) {
+            RelDataType operandType = validator.deriveType(scope, operand);
+            if (!isLiteral(operand) || !isNumeric(operandType)) {
+                continue;
+            }
+            SqlLiteral literal = (SqlLiteral) operand;
+
+            if (literal.getValue() == null) {
+                operandType = TYPE_FACTORY.createSqlType(NULL);
+            } else {
+                BigDecimal numeric = literal.getValueAs(BigDecimal.class);
+                operandType = narrowestTypeFor(numeric, commonType == null ? null : typeName(commonType));
+            }
+
+            commonType = commonType == null ? operandType : withHigherPrecedenceForLiterals(operandType, commonType);
+        }
+
+        // Continue common type inference on non-numeric literals.
+
+        for (SqlNode operand : operands) {
+            RelDataType operandType = validator.deriveType(scope, operand);
+            if (!isLiteral(operand) || isNumeric(operandType)) {
+                continue;
+            }
+            SqlLiteral literal = (SqlLiteral) operand;
+
+            if (literal.getValue() == null) {
+                operandType = TYPE_FACTORY.createSqlType(NULL);
+            } else if (isChar(operandType) && (commonType != null && isNumeric(commonType) || assumeNumeric)) {
+                // Infer proper numeric type for char literals.
+
+                BigDecimal numeric = numericValue(operand);
+                assert numeric != null;
+                operandType = narrowestTypeFor(numeric, commonType == null ? null : typeName(commonType));
+            }
+
+            commonType = commonType == null ? operandType : withHigherPrecedenceForLiterals(operandType, commonType);
+        }
+
+        // seen only parameters
+        if (commonType == null) {
+            assert seenParameters;
+            return null;
+        }
+
+        // can't infer parameter types if seen only NULLs
+        if (typeName(commonType) == NULL && seenParameters) {
+            return null;
+        }
+
+        // fallback to DOUBLE from CHAR, if numeric types assumed
+        if (isChar(commonType) && assumeNumeric) {
+            commonType = TYPE_FACTORY.createSqlType(DOUBLE);
```

Comment:
Why `DOUBLE`, and not `DECIMAL`?

Reply:
The idea was to avoid implicit conversions producing decimals. Still not sure which one to pick, wdyt?

Reply:
The question is - what does it mean for users? Could you please show an example of a query when this branch kicks in? The main problem with DOUBLE is that it is inexact, which may easily lead to surprising results for users.

Reply:
For instance, `SELECT varcharColumn * varcharColumn`.

Reply:
I think, such a combination should not work at all, because it is ambiguous and there is not enough context to understand user's intent.

Reply:
created an issue for 4.2 https://github.com/hazelcast/hazelcast/issues/17311